aten/src/ATen/native/cpu/SpmmReduceKernel.cpp

```diff
@@ -0,0 +1,508 @@
+    for (const auto m : c10::irange(begin, end)) {
+      row_start = csr_data[m];
+      row_end = csr_data[m + 1];
+
+      scalar_t* out_ptr = out_data + m * K;
+
+      constexpr int64_t kVecSize = Vec::size();
+      constexpr int64_t kVLEN = kVecSize * 4;
+      constexpr int64_t CHUNK_SIZE = 16;
+
+      // step 1: reinit the output row for reduce type 'amax' and 'amin'
+      int64_t count = row_end - row_start;
+      if (count != 0) {
+        init<scalar_t, reduce>(out_ptr, K, /*include_self*/false);
+      }
+
+      // step 2: reduce, do blocking on rowwise to reduce write memory bandwidth
+      for (int64_t e0 = row_start; e0 < row_end; e0 += CHUNK_SIZE) {
+        int64_t e1 = std::min(e0 + CHUNK_SIZE, row_end);
+
+        int64_t k = 0;
+        for (; k < K - (K % kVLEN); k += kVLEN) {
+          Vec out_vec0 = Vec::loadu(out_ptr + k);
+          Vec out_vec1 = Vec::loadu(out_ptr + k + kVecSize);
+          Vec out_vec2 = Vec::loadu(out_ptr + k + kVecSize * 2);
+          Vec out_vec3 = Vec::loadu(out_ptr + k + kVecSize * 3);
+          for (const auto e : c10::irange(e0, e1)) {
+            c = col_data[e];
+            scalar_t val = val_data[e];
+            scalar_t* other_ptr = other_data + c * K + k;
+
+            out_vec0 = update<Vec, reduce>(out_vec0, Vec::loadu(other_ptr) * Vec(val));
+            out_vec1 = update<Vec, reduce>(out_vec1, Vec::loadu(other_ptr + kVecSize) * Vec(val));
+            out_vec2 = update<Vec, reduce>(out_vec2, Vec::loadu(other_ptr + kVecSize * 2) * Vec(val));
+            out_vec3 = update<Vec, reduce>(out_vec3, Vec::loadu(other_ptr + kVecSize * 3) * Vec(val));
+          }
+          out_vec0.store(out_ptr + k);
+          out_vec1.store(out_ptr + k + kVecSize);
+          out_vec2.store(out_ptr + k + kVecSize * 2);
+          out_vec3.store(out_ptr + k + kVecSize * 3);
+        }
+        for (; k < K - (K % Vec::size()); k += Vec::size()) {
+          Vec out_vec = Vec::loadu(out_ptr + k);
+          for (const auto e : c10::irange(e0, e1)) {
+            c = col_data[e];
+            scalar_t val = val_data[e];
+            scalar_t* other_ptr = other_data + c * K;
+            out_vec = update<Vec, reduce>(out_vec, Vec::loadu(other_ptr + k) * Vec(val));
+          }
+          out_vec.store(out_ptr + k);
+        }
+        for (; k < K; k++) {
+          scalar_t out_val = out_ptr[k];
+          for (const auto e : c10::irange(e0, e1)) {
+            c = col_data[e];
+            scalar_t val = val_data[e];
+            scalar_t* other_ptr = other_data + c * K;
+            out_val = update<scalar_t, reduce>(out_val, other_ptr[k] * val);
+          }
+          out_ptr[k] = out_val;
+        }
+      }
+
+      // step 3: finalize
+      write<scalar_t, reduce>(out_ptr, count, K);
+    }
+  });
+}
+
+// update both val and arg
+template <typename scalar_t, typename index_t, ReductionType reduce>
+inline void update2(scalar_t *val, scalar_t new_val, index_t *arg, index_t new_arg) {
```

Comment:
Can we provide a better naming of this function or make its usage more clear?

Reply:
updated!